pulsar-transaction/coordinator/src/main/java/org/apache/pulsar/transaction/coordinator/impl/ManagedLedgerTransactionReaderImpl.java

```diff
@@ -0,0 +1,212 @@
+import org.apache.bookkeeper.mledger.impl.PositionImpl;
+import org.apache.pulsar.common.api.proto.PulsarApi.Subscription;
+import org.apache.pulsar.common.api.proto.PulsarApi.TransactionMetadataEntry;
+import org.apache.pulsar.common.api.proto.PulsarApi.TxnStatus;
+import org.apache.pulsar.common.util.protobuf.ByteBufCodedInputStream;
+import org.apache.pulsar.transaction.coordinator.TransactionMetadataStore;
+import org.apache.pulsar.transaction.coordinator.TxnMeta;
+import org.apache.pulsar.transaction.coordinator.TxnSubscription;
+import org.apache.pulsar.transaction.coordinator.exceptions.InvalidTxnStatusException;
+import org.apache.pulsar.transaction.impl.common.TxnID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class ManagedLedgerTransactionReaderImpl implements
+        ManagedLedgerTransactionMetadataStore.ManagedLedgerTransactionReader {
+
+    private static final Logger log = LoggerFactory.getLogger(ManagedLedgerTransactionReaderImpl.class);
+
+    private ConcurrentMap<TxnID, TxnMeta> txnMetaMap = new ConcurrentHashMap<>();
+
+    private AtomicLong sequenceId = new AtomicLong(ManagedLedgerTransactionMetadataStore.TC_ID_NOT_USED);
+
+    private final ReadOnlyCursor readOnlyCursor;
+
+    public ManagedLedgerTransactionReaderImpl(String tcId, ManagedLedgerFactory managedLedgerFactory) throws Exception {
+        this.readOnlyCursor = managedLedgerFactory
+                .openReadOnlyCursor(tcId,
+                        PositionImpl.earliest, new ManagedLedgerConfig());
+    }
+
+    private static List<TxnSubscription> subscriptionToTxnSubscription(List<Subscription> subscriptions) {
+        List<TxnSubscription> txnSubscriptions = new ArrayList<>(subscriptions.size());
+        for (int i = 0; i < subscriptions.size(); i++) {
+            txnSubscriptions
+                    .add(new TxnSubscription(subscriptions.get(i).getTopic(), subscriptions.get(i).getSubscription()));
+        }
+        return txnSubscriptions;
+    }
+
+
+    @Override
+    public TxnMeta getTxnMeta(TxnID txnID) {
+        return txnMetaMap.get(txnID);
+    }
+
+    @Override
+    public Long readSequenceId() {
+        return sequenceId.get();
+    }
+
+    @Override
+    public void addNewTxn(TxnMeta txnMeta) {
+        txnMetaMap.put(txnMeta.id(), txnMeta);
+    }
+
+    @Override
+    public TxnStatus getTxnStatus(TxnID txnID) {
+        return txnMetaMap.get(txnID).status();
+    }
+
+    @Override
+    public CompletableFuture<Void> init(TransactionMetadataStore transactionMetadataStore) {
+        CountDownLatch countDownLatch = new CountDownLatch(1);
+        countDownLatch.countDown();
+        CompletableFuture<Void> completableFuture = new CompletableFuture<>();
+        transactionMetadataStore.updateMetadataStoreState(TransactionMetadataStore.State.INITIALIZING);
+        readOnlyCursor
+                .asyncReadEntries(100,
+                        new ReaderReadEntriesCallback(countDownLatch, txnMetaMap, sequenceId,
```

Comment:
every 100 messages will create a ReaderReadEntriesCallback instance, i think here can refer to approach like consumer read messages or `PulsarRecordCursor` in Pulsar SQL.

Reply:
Also, expose configs for each read batch size.